cleandart/react-dart
# Improve management of Dart component instances

___Note: for optimal experience when viewing the diff, [use the whitespace-ignoring flag: `?w=1`](https://github.com/cleandart/react-dart/pull/127/files?w=1).___

# Problem
Dart component instances were previously tied to the JS props and the Dart `ReactDartComponentInternal` objects.

 This led to confusing lifecycle of the Dart components, as well as issues when attempting to clean up leaks in that lifecycle (see https://github.com/cleandart/react-dart/pull/125).
> Unmounted component instances were retained by the original `ReactElement`'s props via the `ReactDartComponentInternal` object, which is used to tie the Dart component instance to the JS component instance.

# Changes
- Improve management of Dart component instances by storing them on the JS instances. This ensures that the lifecycle of Dart instances is in sync with the JS instances.
    - __Breaking change:__ Remove `ReactDartComponentInternal.component` - obviated by new JS instance storage
    - __Breaking change:__ Remove `ReactDartComponentInternal.isMounted` - not used, wasn't accurate after unmount


#### Other changes
- Add script to concatenate React and custom JS files sourced from `js_src` directory
    - Did this mostly for my own sanity to ensure I didn't mess anything up when copying custom JS from file to file
    - This resulted in a large number of whitespace changes, which can be ignored in this PR (see instructions at top about whitespace-agnostic diff viewing)
- Add transformer so tests can be run in DDC using `--pub-serve` flag

# Testing
- Run tests in content-shell, dart2js, and DDC locally to ensure that everything works as expected
- Verify that the memory leak is fixed:
    - Check out test branch: https://github.com/greglittlefield-wf/react-dart/tree/internal_instance_refactor-leak_test
    - Load `example/leak.html` with the dev tools open and following the instructions in the console
- Verify tests pass in over_react and WSD test branches:
    - https://github.com/Workiva/over_react/compare/master...greglittlefield-wf:internal_instance_refactor
    - https://github.com/Workiva/web_skin_dart/compare/master...greglittlefield-wf:internal_instance_refactor

# Areas of Regression
Component lifecycle
